src/Lean/Elab/Tactic/ElabTerm.lean

```diff
@@ -372,10 +372,23 @@ private def preprocessPropToDecide (expectedType : Expr) : TermElabM Expr := do
     let expectedType ← preprocessPropToDecide expectedType
     let d ← mkDecide expectedType
     let d ← instantiateMVars d
-    let r ← withDefault <| whnf d
-    unless r.isConstOf ``true do
-      throwError "failed to reduce to 'true'{indentExpr r}"
-    let s := d.appArg! -- get instance from `d`
+    -- Get instance from `d`
+    let s := d.appArg!
+    -- Reduce the instance rather than `d` itself, since that gives a nicer error message on failure.
+    let r ← withDefault <| whnf s
+    if r.isAppOf ``isFalse then
+      throwError "\
+        tactic 'decide' proved that the proposition\
+        {indentExpr expectedType}\n\
+        is false"
+    unless r.isAppOf ``isTrue do
+      throwError "\
+        tactic 'decide' failed for proposition\
+        {indentExpr expectedType}\n\
+        since its 'Decidable' instance did not reduce to the 'isTrue' constructor:\
+        {indentExpr r}"
```

Comment:
I find this phrasing ever so slightly confusing - what about this?
```suggestion
        since its 'Decidable' instance did not reduce to the 'isTrue' constructor.\n\
        It reduced to:\
        {indentExpr r}"
```
Otherwise it sounds to me like the expression is the constructor, instead of what we got instead.